### PR TITLE
feat: add TGUI for stack crafting

### PR DESCRIPTION
**Pull request by Gaxeer**
## What Does This PR Do
Adds TGUI for stack crafting (crafts from different sheets e.g metal, glass etc.)
Mostly copy of https://github.com/tgstation/tgstation/pull/53657

## Why It's Good For The Game
Pretty and useful UI with search

## Images of changes
<details>
<summary>Video of UI testing</summary>

https://github.com/ParadiseSS13/Paradise/assets/44334376/f0e565fa-7afd-4c6c-af4d-a7766982fdb7

</details>

## Changelog
:cl:
add: TGUI for stack crafting (metal sheets etc.)
/:cl:

### CODE REVIEW

File: code/game/objects/items/stacks/stack.dm [hunk 1]
```diff
@@ -124,169 +122,56 @@
...
-				message_admins("[key_name_admin(usr)] just attempted to href exploit sheet crafting with an invalid multiplier. Ban highly advised.")
-			multiplier = 1
-
-		if(get_amount() < R.req_amount * multiplier)
-			if(R.req_amount * multiplier > 1)
-				to_chat(usr, "<span class='warning'>You haven't got enough [src] to build \the [R.req_amount * multiplier] [R.title]\s!</span>")
-			else
-				to_chat(usr, "<span class='warning'>You haven't got enough [src] to build \the [R.title]!</span>")
-			return FALSE
+		else if (istype(recipe, /datum/stack_recipe_list))
```
Comment (DGamerL) [inline code/game/objects/items/stacks/stack.dm:160]:
```suggestion
		else if(istype(recipe, /datum/stack_recipe_list))
```

File: code/game/objects/items/stacks/stack.dm [hunk 2]
```diff
@@ -124,169 +122,56 @@
 	else
 		..()
 
+/obj/item/stack/ui_interact(mob/user, ui_key, datum/tgui/ui, force_open, datum/tgui/master_ui, datum/ui_state/state = GLOB.hands_state)
+	ui = SStgui.try_update_ui(user, src, ui_key, ui, force_open)
+	if(!ui)
+		ui = new(user, src, ui_key, "StackCraft", name, 400, 400, master_ui, state)
+		ui.set_autoupdate(TRUE)
```
Comment (GDNgit) [inline code/game/objects/items/stacks/stack.dm:129]: good lord please do not have this autoupdate

**Comment by S34NW** [comment]
Nice, will complete https://github.com/orgs/ParadiseSS13/projects/4/views/1?filterQuery=craf&pane=issue&itemId=38641824

**Comment by DogeDogIs** [comment]
Would be cool if you can put some images, for new players this would be great

**Comment by Gaxeer** [comment]
> Would be cool if you can put some images, for new players this would be great

ye, why not. Will try to do it

File: code/game/objects/items/stacks/stack.dm [hunk 3]
```diff
@@ -301,6 +184,85 @@
...
+	if(get_amount() < 1 && !is_cyborg)
+		qdel(src)
+		return FALSE
+
+	if(!validate_build(user, recipe_to_make, multiplier))
+		return FALSE
+
+	if(recipe_to_make.time)
+		to_chat(usr, "<span class='notice'>Building [recipe_to_make.title]...</span>")
+		if(!do_after(usr, recipe_to_make.time, target = user))
```
Comment (Burzah) [inline code/game/objects/items/stacks/stack.dm:197]:
```suggestion
		to_chat(user, "<span class='notice'>Building [recipe_to_make.title]...</span>")
		if(!do_after(user, recipe_to_make.time, target = user))
```
I believe these can be changed to `user`.

**Review by S34NW — CHANGES_REQUESTED** [review]
TGUI update is in, please deconflict and update this.

**Comment by S34NW** [comment]
No change in over a week, please feel free to re-open or make a new PR once you find the time to tackle this

**Pull request closed without merging**